From 711a5872536bddd463ad43b6131c93295b2ae14e Mon Sep 17 00:00:00 2001
From: Wenzong Fan <wenzong.fan@windriver.com>
Date: Fri, 4 Mar 2016 01:52:02 -0500
Subject: [PATCH] wr-policy: allow initrc_t start systemd_unit_t service

Reference sources: selinux-policy-3.13.1-128.21.fc22.src.rpm
Changes come from: policy-f22-base.patch

* Fix label for /sbin/sulogin.util-linux
* Allow initrc_t start, reload systemd_unit_t services
* Add initial local policy for systemd_passwd_agent_t
* Allow sulogin_t uses console, ttys, ptys
* Allow sulogin_t getattr from chr_files, blk_files
* Allow sulogin_t read generic crypto sysctls

Fix avc denials:

  avc: denied { transition } for pid=1023 comm="sulogin" \
  path="/bin/bash" dev="hda" ino=27481 \
  scontext=system_u:system_r:initrc_t:s0-s15:c0.c1023 \
  tcontext=root:sysadm_r:sysadm_t:s0 \
  tclass=process permissive=1

  avc: denied { search } for pid=1075 comm="sulogin" \
  name="sys" dev="proc" ino=4026531855 \
  scontext=system_u:system_r:sulogin_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:sysctl_t:s0 \
  tclass=dir permissive=0

  avc: denied { getattr } for pid=985 comm="sulogin" \
  path="/dev/fuse" dev="devtmpfs" ino=8914 \
  scontext=system_u:system_r:sulogin_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:fuse_device_t:s0 \
  tclass=chr_file permissive=1

  avc: denied { getattr } for pid=985 comm="sulogin" \
  path="/dev/vcsa6" dev="devtmpfs" ino=14541 \
  scontext=system_u:system_r:sulogin_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:tty_device_t:s0 \
  tclass=chr_file permissive=1

  avc: denied { getattr } for pid=985 comm="sulogin" \
  path="/dev/ppp" dev="devtmpfs" ino=8918 \
  scontext=system_u:system_r:sulogin_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:ppp_device_t:s0 \
  tclass=chr_file permissive=1

  avc: denied { start } for auid=n/a uid=0 gid=0 \
  path="/lib/systemd/system/multi-user.target" \
  cmdline="/bin/systemctl --job-mode=fail --no-block default" \
  scontext=system_u:system_r:initrc_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:systemd_unit_t:s0 \
  tclass=service

  avc: denied { read write } for pid=984 comm="systemd-tty-ask" \
  path="/dev/console" dev="devtmpfs" ino=7265 \
  scontext=system_u:system_r:systemd_passwd_agent_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:console_device_t:s0 \
  tclass=chr_file permissive=0

  avc: denied { setfscreate } for pid=940 comm="systemd-tty-ask" \
  scontext=system_u:system_r:systemd_passwd_agent_t:s0-s15:c0.c1023 \
  tcontext=system_u:system_r:systemd_passwd_agent_t:s0-s15:c0.c1023 \
  tclass=process permissive=1

  avc: denied { read } for pid=993 comm="systemd-tty-ask" \
  name="ask-password" dev="tmpfs" ino=8192 \
  scontext=system_u:system_r:systemd_passwd_agent_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:init_var_run_t:s0 \
  tclass=dir permissive=1

Upstream-Status: Pending

Signed-off-by: Wenzong Fan <wenzong.fan@windriver.com>
---
 policy/modules/kernel/devices.te    |  1 +
 policy/modules/system/init.if       | 37 +++++++++++++++++++++++++++++++++++++
 policy/modules/system/init.te       |  2 ++
 policy/modules/system/locallogin.fc |  1 +
 policy/modules/system/locallogin.te | 10 ++++++++++
 policy/modules/system/systemd.te    | 18 ++++++++++++++++++
 6 files changed, 69 insertions(+)

diff --git a/policy/modules/kernel/devices.te b/policy/modules/kernel/devices.te
index 37dceb4..8585526 100644
--- a/policy/modules/kernel/devices.te
+++ b/policy/modules/kernel/devices.te
@@ -21,6 +21,7 @@ files_mountpoint(device_t)
 files_associate_tmp(device_t)
 fs_xattr_type(device_t)
 fs_use_trans devtmpfs gen_context(system_u:object_r:device_t,s0);
+dev_node(device_t)
 
 #
 # Type for /dev/agpgart
diff --git a/policy/modules/system/init.if b/policy/modules/system/init.if
index 90b2480..c624a68 100644
--- a/policy/modules/system/init.if
+++ b/policy/modules/system/init.if
@@ -2543,6 +2543,43 @@ interface(`init_stop_all_units',`
 	allow $1 systemdunit:service stop;
 ')
 
+########################################
+## <summary>
+##      Read/Write init unnamed pipes.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      Domain allowed access.
+##      </summary>
+## </param>
+#
+interface(`init_rw_pipes',`
+	gen_require(`
+		type init_var_run_t;
+	')
+
+	rw_fifo_files_pattern($1, init_var_run_t, init_var_run_t)
+')
+
+#######################################
+## <summary>
+##  Create a directory in the /run/systemd directory.
+## </summary>
+## <param name="domain">
+##  <summary>
+##  Domain allowed access.
+##  </summary>
+## </param>
+#
+interface(`init_create_pid_dirs',`
+	gen_require(`
+		type init_var_run_t;
+	')
+
+	allow $1 init_var_run_t:dir list_dir_perms;
+	create_dirs_pattern($1, init_var_run_t, init_var_run_t)
+')
+
 #######################################
 ## <summary>
 ##	Reload all systemd units.
diff --git a/policy/modules/system/init.te b/policy/modules/system/init.te
index 5004c17..20b9b14 100644
--- a/policy/modules/system/init.te
+++ b/policy/modules/system/init.te
@@ -784,6 +784,8 @@ ifdef(`init_systemd',`
 	init_rw_stream_sockets(initrc_t)
 	init_get_all_units_status(initrc_t)
 	init_stop_all_units(initrc_t)
+	init_reload_all_units(initrc_t)
+	init_start_all_units(initrc_t)
 
 	# Create /etc/audit.rules.prev after firstboot remediation
 	logging_manage_audit_config(initrc_t)
diff --git a/policy/modules/system/locallogin.fc b/policy/modules/system/locallogin.fc
index 06b5de6..9023860 100644
--- a/policy/modules/system/locallogin.fc
+++ b/policy/modules/system/locallogin.fc
@@ -1,5 +1,6 @@
 
 /sbin/sulogin		--	gen_context(system_u:object_r:sulogin_exec_t,s0)
+/sbin/sulogin.util-linux	--	gen_context(system_u:object_r:sulogin_exec_t,s0)
 /sbin/sushell		--	gen_context(system_u:object_r:sulogin_exec_t,s0)
 
 /usr/sbin/sulogin	--	gen_context(system_u:object_r:sulogin_exec_t,s0)
diff --git a/policy/modules/system/locallogin.te b/policy/modules/system/locallogin.te
index 8e35567..976a595 100644
--- a/policy/modules/system/locallogin.te
+++ b/policy/modules/system/locallogin.te
@@ -227,6 +227,7 @@ allow sulogin_t self:msgq create_msgq_perms;
 allow sulogin_t self:msg { send receive };
 
 kernel_read_system_state(sulogin_t)
+kernel_read_crypto_sysctls(sulogin_t)
 # because file systems are not mounted:
 kernel_dontaudit_search_unlabeled(sulogin_t)
 
@@ -268,6 +269,15 @@ ifdef(`sulogin_no_pam', `
 	selinux_compute_user_contexts(sulogin_t)
 ')
 
+dev_getattr_all_chr_files(sulogin_t)
+dev_getattr_all_blk_files(sulogin_t)
+
+mls_file_read_all_levels(sulogin_t)
+
+term_use_console(sulogin_t)
+term_use_generic_ptys(sulogin_t)
+term_use_all_ttys(sulogin_t)
+
 optional_policy(`
 	nis_use_ypbind(sulogin_t)
 ')
diff --git a/policy/modules/system/systemd.te b/policy/modules/system/systemd.te
index 1246875..927ec3a 100644
--- a/policy/modules/system/systemd.te
+++ b/policy/modules/system/systemd.te
@@ -420,3 +420,21 @@ tunable_policy(`systemd_tmpfiles_manage_all',`
 	files_relabel_non_security_dirs(systemd_tmpfiles_t)
 	files_relabel_non_security_files(systemd_tmpfiles_t)
 ')
+
+#######################################
+#
+# Systemd_passwd_agent local policy
+#
+
+allow systemd_passwd_agent_t self:capability { chown sys_tty_config dac_override };
+allow systemd_passwd_agent_t self:process { setfscreate signal_perms setsockcreate };
+allow systemd_passwd_agent_t self:unix_dgram_socket create_socket_perms;
+
+term_read_console(systemd_passwd_agent_t)
+term_write_console(systemd_passwd_agent_t)
+
+init_search_run(systemd_passwd_agent_t)
+init_rw_pipes(systemd_passwd_agent_t)
+init_create_pid_dirs(systemd_passwd_agent_t)
+init_read_utmp(systemd_passwd_agent_t)
+init_stream_connect(systemd_passwd_agent_t)
-- 
2.7.4

